File: backend/app/schema/ProfileAnalytics.py
from uuid import UUID
from datetime import datetime
from pydantic import BaseModel, ConfigDict


class UserStackStatResponse(BaseModel):
    id: UUID
    stack_name: str
    projects_count: int
    live_projects_count: int
    journal_entries_count: int
    score: int
    level: int
    last_used_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class ProfileProjectResponse(BaseModel):
    id: UUID
    title: str
    slug: str
    description: str | None
    github_url: str | None
    live_url: str | None
    thumbnail_url: str | None
    tech_stack: list[str]
    stars_count: int
    views_count: int
    comments_count: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ProfileLiveProjectResponse(BaseModel):
    id: UUID
    title: str
    slug: str
    goal: str
    progress_percentage: int
    status: str
    tech_stack: list[str]
    journal_count: int
    views_count: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class UserFullProfileResponse(BaseModel):
    id: UUID
    username: str
    clerk_user_id: str
    display_name: str | None
    email: str
    bio: str | None
    avatar_url: str | None
    banner_url: str | None
    github_url: str | None
    linkedin_url: str | None
    portfolio_url: str | None
    location: str | None
    current_build: str | None

    reputation_score: int
    followers_count: int
    following_count: int
    posts_count: int
    project_count: int

    projects: list[ProfileProjectResponse]
    live_projects: list[ProfileLiveProjectResponse]
    stack_stats: list[UserStackStatResponse]

    model_config = ConfigDict(from_attributes=True)

File: frontend/app/lib/type/profileAnalytics.ts
// =========================================================
// STACK STATS
// =========================================================

export type UserStackStat = {
    id: string;

    stack_name: string;

    projects_count: number;

    live_projects_count: number;

    journal_entries_count: number;

    score: number;

    level: number;

    last_used_at: string | null;
};


// =========================================================
// PROJECT
// =========================================================

export type ProfileProject = {
    id: string;

    title: string;

    slug: string;

    description: string | null;

    github_url: string | null;

    live_url: string | null;

    thumbnail_url: string | null;

    tech_stack: string[];

    stars_count: number;

    views_count: number;

    comments_count: number;

    created_at: string;
};


// =========================================================
// LIVE PROJECT
// =========================================================

export type ProfileLiveProject = {
    current_goal: string;
    id: string;

    title: string;

    slug: string;

    goal: string;

    progress_percentage: number;

    status: string;

    tech_stack: string[];

    journal_count: number;

    views_count: number;

    created_at: string;
};


// =========================================================
// FULL PROFILE
// =========================================================

export type UserFullProfile = {
    id: string;

    clerk_user_id: string;

    username: string;

    display_name: string | null;

    email: string;

    bio: string | null;

    avatar_url: string | null;

    banner_url: string | null;

    github_url: string | null;

    linkedin_url: string | null;

    portfolio_url: string | null;

    instagram_url: string | null;

    location: string | null;

    current_build: string | null;

    joined_date: string;

    reputation_score: number;

    followers_count: number;

    following_count: number;

    posts_count: number;

    project_count: number;

    projects: ProfileProject[];

    live_projects: ProfileLiveProject[];

    stack_stats: UserStackStat[];
};